plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,300 @@
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting
+ * top N queries with high latency data for search queries
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsService {
+    private static final Logger log = LogManager.getLogger(QueryInsightsService.class);
+
+    private static final TimeValue delay = TimeValue.ZERO;
+    /**
+     * The internal OpenSearch thread pool that execute async processing and exporting tasks
+     */
+    private final ThreadPool threadPool;
+
+    /**
+     * enable insight data collection
+     */
+    private final Map<MetricType, Boolean> enableCollect = new HashMap<>();
+
+    private int topNSize = QueryInsightsSettings.DEFAULT_TOP_N_SIZE;
+
+    private TimeValue windowSize = TimeValue.timeValueSeconds(QueryInsightsSettings.DEFAULT_WINDOW_SIZE);
+    /**
+     * The internal thread-safe store that holds the top n queries insight data, by different MetricType
+     */
+    private final Map<MetricType, PriorityBlockingQueue<SearchQueryRecord>> topQueriesStores;
+
+    /**
+     * The internal store that holds historical top n queries insight data by different MetricType in the last window
+     */
+    private final Map<MetricType, List<SearchQueryRecord>> topQueriesHistoryStores;
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart = new HashMap<>();
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        topQueriesStores = new HashMap<>();
+        topQueriesHistoryStores = new HashMap<>();
+        for (MetricType metricType : MetricType.allMetricTypes()) {
+            topQueriesStores.put(metricType, new PriorityBlockingQueue<>(topNSize, (a, b) -> SearchQueryRecord.compare(a, b, metricType)));
+            topQueriesHistoryStores.put(metricType, new ArrayList<>());
+            topQueriesWindowStart.put(metricType, -1L);
+        }
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Ingest the query data into local stores
+     *
+     * @param record the record to ingest
+     */
+    public void addRecord(SearchQueryRecord record) {
```

Comment:
Is this function thread safe?

Reply:
We are using PriorityBlockingQueue<SearchQueryRecord> so it should be thread safe.